Get started fixes

zemccartney opened the pull request:
Addresses https://github.com/hapipal/hapipal.com/issues/50

Relative to the list on that issue:
1. ✅ 
2. ✅ 
3. Updated at one point in the tutorial, appears most links were in place
4. ✅ 
5. ✅ 

Additionally,

- removed explanation of `npx`, given that tool is more commonplace (it was brand new at the time of original writing)
- added boilerplate's git project init command to top of the tutorial, to head off issues with applying flavors failing due to overwriting local changes
- updated route config method capitalization to match convention in example repo
- removed a handful of usages of the word "just"; realized it was brutally overused (to be clear, by me, my bad writing habit; to my ear at least 😛  )

Lastly, to the issue's suggestion of opening links in new tabs, this was a bit tricky: setting `target=_blank` within the markdown file didn't work; the attribute was stripped by Github's markdown API, which we use to process the file to HTML.

To workaround this, I experimented with 2 strategies, neither of which struck me as great, but both work:

- **pre-process**: as seen in `lib/routes/getting-started.js`, prior to passing the html received from Github off to handlebars, we process it by adding `target="_blank"` to all `a` tags
- **handlebars**: we lean into our usage of Handlebars by
    - use handlebars expressions to represent links e.g. `{{link "here" "github.com/hapipal/examples/tree/master/paldo-riddles" true}}`
        - the protocol is separated because otherwise, github translates the valid URL to an anchor element with an href attribute, on which the `link` helper chokes
    - handle these expressions with the `link` helper
    - compile these expressions within the html processed by Github via the `precomp` helper, which compiles the html before passing it off to our existing template to set as a normal expression; without this step, the expressions appear literally. Concretely, in `getting-started.hbs`, replace `{{{html}}}` with `{{{precomp html true }}}`

@devinivy some questions for ya:

- re: opening links in new tabs, are you cool with either strategy above? Or should I cut that work completely, cool to leave links as-is?
- Should I update all hapi documentation link in the tutorial (and elsewhere on the site) to hapi.dev, not API.md on github? Not crucial I know, but thinking to be consistent with how each repo directs users to hapi.dev
- I was thinking adding a seed, to create list of riddles, instead of depending on people to manually recreate their own could smooth out testing the later steps in the tutorial; how does that sound to you? I realized on reading back through that having to manually recreate riddles on server startup could annoy, if not trip people up. Another non-critical, thinking just as a nice-to-have

zemccartney commented on lib/services/github.js:
```diff
@@ -1,6 +1,7 @@
 'use strict';
 
 const { promises: Fs } = require('fs');
+const Stream = require('stream');
```
```suggestion
```

devinivy commented on lib/content/getting-started.md:
```diff
@@ -38,17 +38,17 @@ First things first, we need to setup a base pal project. Run the following:
- - execute package commands installed locally to your project without specifying the path to the package
-
-<sup>[There's more to npx](https://medium.com/@maybekatz/introducing-npx-an-npm-package-runner-55f7d4bd282b), if you're curious.</sup>
+`npx hpal new paldo-riddles` calls hapi pal's command line utility [hpal](https://www.npmjs.com/package/hpal) to bootstrap a new project in a directory titled `paldo-riddles` in our current working directory (the argument to `new` is a path).
```
```suggestion
`npx hpal new paldo-riddles` calls hapi pal's command line utility [hpal](https://www.npmjs.com/package/@hapipal/hpal) to bootstrap a new project in a directory titled `paldo-riddles` in our current working directory (the argument to `new` is a path).
```

devinivy commented on lib/content/getting-started.md:
```diff
@@ -38,17 +38,17 @@ First things first, we need to setup a base pal project. Run the following:
 npx hpal new paldo-riddles
```
```suggestion
npm init @hapipal paldo-riddles
```

devinivy commented on lib/content/getting-started.md:
```diff
@@ -38,17 +38,17 @@ First things first, we need to setup a base pal project. Run the following:
 cd paldo-riddles
 npm install
+
+# make your first commit to init project history
+git add --all
```
I believe all the files are already added after starting a new project.  However, after installation there is one new file, package-lock.json, that this would add.

devinivy replied:
I see now that this mirrors the boilerplate readme.  Maybe it is easier to leave like this than get granular about adding specific files 👍

devinivy commented on lib/content/getting-started.md:
```diff
@@ -232,7 +222,7 @@ module.exports = {
 Be sure to restart your server in order to pick-up this new code.
 
-If you cURL our new route (`curl http://localhost:3000/riddle-random`) or visit it in your browser, you'll see one of Paldo's riddles. We're up and running!
+If you cURL our new route (`curl http://localhost:3000/riddle-random`) or visit it [in your browser](http://localhost:3000/riddle-random), you'll see one of Paldo's riddles. We're up and running!
```
Above there were some instructions removed to set the PORT back to 3000.  Does that mean we should use 4000 throughout the rest of the tutorial?

zemccartney replied:
I can't for the life of me recall what I was thinking there. I'll add some note about the rest of the tutorial serving on 3000 after that env setting section

zemccartney replied:
H'ok, added back a note explaining this away!

devinivy commented on lib/content/getting-started.md:
```diff
@@ -248,7 +238,7 @@ Moving on!
 Immediately, we see that our strategy of hardcoding our riddles within our first route's handler is, although expedient, unworkable. Our other routes will need to know about that data (let alone any other pieces of our riddle-sharing application we build later). So, let's centralize it.
 
-There's a whole slew of ways you handle this. A simple one: we just create a file called `data` under `lib` and set it up to export our riddles.
+Though there's a whole slew of ways you handle this, we'll create a file called `data` under `lib` and set it up to export our riddles.
```
I wonder if it might be a little more comfy if we don't mention that there are tons of ways to solve the problem, and just go ahead with a solution that will work for us.
```suggestion
For our purposes, we'll create a file called `data` under `lib` and set it up to export our riddles.
```

zemccartney replied:
Ay, yea, that reads much better. Thanks!

devinivy commented on lib/content/getting-started.md:
```diff
@@ -342,7 +332,7 @@ We can take advantage of this in a couple of ways.
 ### Manually
 
-Just run `npm run lint`—this executes eslint with our config on all files not ignored by npm (see the project's standard `.npmignore`)—and then fix whatever warnings and errors it spits out.
+Run `npm run lint`—this executes eslint with our config on all files not ignored by npm (see the project's standard `.npmignore`)—and then fix whatever warnings and errors it spits out (or if you prefer eslint's automatic fixing: `npm run lint -- --fix`).
```
Ooh, cool tip!

devinivy commented on lib/content/getting-started.md:
````diff
@@ -372,21 +362,20 @@ Objection ORM is an impressive SQL query-builder with a fantastic community, bui
 
 ```sh
 git cherry-pick objection
-npm install
````
Is there an instruction elsewhere to install the new packages (objection, knex, etc.)?

zemccartney replied:
Ay, yea, it's a bunch of lines down, after explaining what all's happening with this command: https://github.com/hapipal/hapipal.com/pull/52/files#diff-663e1a280a7fe30ead22732740aa3d215335bf0c66967a35595d2bf153ae2108L435

Does that read alright to you? I was kept that as the only place where we instruct to install since the instructions right after this allude to resolving merge conflicts in package.json. But I could see lifting those up, so it reads something like: 

```diff
Expect to resolve small merge conflicts when pulling flavors in, typically just in `package.json`  and `server/manifest.js` having to do with overlapping dependencies in HEAD and the flavor.

+ Now let's check everything's still working:
+ 
+ ```sh
+ # bring in our new dependencies
+ npm install
+ npm start
+ ```
```

Do you prefer one over the other?

devinivy replied:
Sounds like it's spoken to, all good to leave it as-is 👍

devinivy commented on lib/content/getting-started.md:
```diff
@@ -782,7 +776,7 @@ Hey, this is a pretty good start for Paldo—good work!  As you can see, there's
 - [hpal](https://hapipal.com/docs/hpal) - this is the command line tool we used to start a new project, create routes in `routes/`, and models in `models/`.  It does much more too—you can also search documentation with it from the command line, for example: `hpal docs:schwifty request.models`.
 - [haute-couture](https://hapipal.com/docs/haute-couture) - this is used by the pal boilerplate to enforce the directory structure for your hapi plugin (everything in `lib/`).
-- [joi](https://github.com/hapijs/joi/blob/master/API.md) - this is the validation library of choice for hapi projects, since it integrates nicely into hapi itself.
+- [joi](https://github.com/sideway) - this is the validation library of choice for hapi projects, since it integrates nicely into hapi itself.
```
```suggestion
- [joi](https://github.com/sideway/joi) - this is the validation library of choice for hapi projects, since it integrates nicely into hapi itself.
```

devinivy commented:
Ay!  Thanks for this, looks great and it will feel so good to resolve these.

> re: opening links in new tabs, are you cool with either strategy above? Or should I cut that work completely, cool to leave links as-is?

I think we'll be stuck with a hack no matter what, since as you mentioned there's no support for this in GitHub markdown.  My preference would actually be some post-processing on the frontend since I think it will be the most straightforward to understand and maintain.  Something to this effect:
```js
const markdownLinksExternal = () => {

    const links = document.querySelectorAll('.markdown-body a');

    for (let i = 0; i < links.length; ++i) {
        const link = links[i];
        link.setAttribute('target', '_blank');
    }
};
```

> Should I update all hapi documentation link in the tutorial (and elsewhere on the site) to hapi.dev, not API.md on github? Not crucial I know, but thinking to be consistent with how each repo directs users to hapi.dev

If you are open to taking the time to do that, that would be awesome!

> I was thinking adding a seed, to create list of riddles, instead of depending on people to manually recreate their own could smooth out testing the later steps in the tutorial; how does that sound to you? I realized on reading back through that having to manually recreate riddles on server startup could annoy, if not trip people up. Another non-critical, thinking just as a nice-to-have

I'm down with that, but if it would be simpler for the purposes of the tutorial to use a file-backed SQLite persistence then that's cool too.

We can also pull any of these out into separate issues and address later if you'd like 👍

zemccartney commented:
Alright, I _think_ I addressed everything (thanks for all those detailed notes, btw! 🙏 )

- added link target post-processing, cleaned up prior junk
- switched all links to hapi and joi docs to point at their respective sites instead of github
- wrote up instructions for persisting data to a sqlite file (and pointed at the premade one in `examples`)
- tidied up the other suggestions

Though let me know if there's anything else that'd be good to address while I'm in here!

devinivy commented on lib/public/js/main.js:
```diff
@@ -194,8 +194,20 @@ const palToClipboard = () => {
+const markdownLinksExternal = () => {
+
+    // external links only (no section anchors or internal (#) links)
+    const links = document.querySelectorAll('.markdown-body.entry-content a:not([href^="#"]):not(.anchor)');
```
😏

zemccartney replied:
🎩 
🍹

devinivy reviewed (approved):
Beautiful, thanks again 🙏
